client/network/src/protocol/notifications/handler.rs

```diff
@@ -159,6 +159,16 @@ enum State {
 	Closed {
 		/// True if an outgoing substream is still in the process of being opened.
 		pending_opening: bool,
+
+		/// Outbound substream that has been accepted by the remote. Being closed.
+		out_substream_closing: Option<NotificationsOutSubstream<NegotiatedSubstream>>,
```

Comment:
Ok, I guess the other questions are rather similar. I think I need some context of this state machine to make sense of the code.

Reply:
The enum variant `Closed` represents the state the state machine exposes to the outside (i.e. the behaviour).
When the behaviour asks for closing, we instantaneously respond "ok" and switch to `Closed`. There is "closing" state internally, but we keep the existing substreams in order to close them properly.

There is a delicate situation to handle: the behaviour might want to close, then open, then close, then open, then close, then open, etc. and the handler would need to respond to all these requests one by one.
In order to avoid such a situation, it is illegal for the behaviour to try open when a previous open request hasn't been confirmed yet. However, in order to not make the implementation of the behaviour too complex, it is always legal to close whatever the situation.
This is implemented in the handler by immediately responding to all close requests.

In other words, the API is designed in order to spread the implementation complexity between the behaviour and the handler.

Reply:
Thanks Pierre, that helps!